### ISSUE TITLE
Miscompile with large structs passed on the stack

### ISSUE THREAD

**Issue by LiterallyVoid** [ziglang/zig#14706]
### Zig Version

0.11.0-dev.1782+b52be973d

### Environment

Linux x86_64, with LLVM version 15.0.7.

### Steps to Reproduce and Observed Behavior

Create a file containing these contents:

```zig
const std = @import("std");

const Thing = extern struct {
    first: i32,
    big: u128 = 0,
};

pub noinline fn a(thing: Thing) callconv(.C) i32 {
    std.log.err("in a, got {}", .{ thing });
    return thing.first;
}

pub noinline fn b(thing: Thing) callconv(.C) i32 {
    return a(thing);
}

pub noinline fn createThing(inner: i32) Thing {
    return Thing {
        .first = inner,
    };
}

test "test" {
    std.log.err("{}", .{ b(createThing(0)) });
    std.log.err("{}", .{ b(createThing(1)) });
}
```

Run the command `zig test code.zig -O ReleaseFast`.

Actual output with any optimizations enabled:
```
$ zig test a.zig -O ReleaseFast
Test [1/1] test.test... [default] (err): in a, got a.Thing{ .first = 0, .big = 0 }
[default] (err): 0
[default] (err): in a, got a.Thing{ .first = 208512, .big = 11982861047231205721541652096 }
[default] (err): 208512
All 1 tests passed.
4 errors were logged.
error: the following test command failed with exit code 1:
/home/§/.cache/zig/o/1b43972b32f88a3250693b59ce6779ab/test
```

(The numbers are all random data from the stack.)

### Expected Behavior

Desired output:

```
$ zig test a.zig -O Debug
Test [1/1] test.test... [default] (err): in a, got a.Thing{ .first = 0, .big = 0 }
[default] (err): 0
[default] (err): in a, got a.Thing{ .first = 1, .big = 0 }
[default] (err): 1
All 1 tests passed.
4 errors were logged.
error: the following test command failed with exit code 1:
/home/§/.cache/zig/o/65cf4b9db6aaf3930b574f006def2dc2/test
```

**Comment by LiterallyVoid** [comment]
I’ve already found the problem.

If structures are required to be passed on the stack, Zig puts `byval` in the LLVM function signature.

Here's what the signature of the `a` function looks like:
```llvm
define internal i32 @a.a(ptr nonnull readonly byval(%a.Thing) align 16 %0) unnamed_addr #1 !dbg !2825 {
```

And when it's called from the `b` function, Zig generates the following LLVM IR:
```llvm
  %2 = call i32 @a.a(ptr %0), !dbg !2600
```

This is the wrong IR! `byval` is part of the ABI, and this mismatch specifically will create a miscompile that is incredibly hard to reproduce.
The correct IR would look something like this:
```llvm
  %2 = call i32 @a.a(ptr byval(%a.Thing) %0), !dbg !2600
```

[Here's the relevant LLVM documentation](https://llvm.org/docs/LangRef.html#id328):

> […] if the calling convention is not swifttailcc or tailcc:
>
> * All ABI-impacting function attributes, such as sret, byval, inreg, returned, and inalloca, must match.
> * …

**Comment by Vexu** [comment]
Adding attributes to calls does seem to fix this:
```diff
diff --git a/src/codegen/llvm.zig b/src/codegen/llvm.zig
index aa794827a..be1827d16 100644
--- a/src/codegen/llvm.zig
+++ b/src/codegen/llvm.zig
@@ -4915,8 +4915,6 @@ pub const FuncGen = struct {
             "",
         );
 
-        if (callee_ty.zigTypeTag() == .Pointer) {
-            // Add argument attributes for function pointer calls.
             it = iterateParamTypes(self.dg, fn_info);
             it.llvm_index += @boolToInt(sret);
             it.llvm_index += @boolToInt(err_return_tracing);
@@ -4973,7 +4971,6 @@ pub const FuncGen = struct {
                     }
                 },
             };
-        }
 
         if (return_type.isNoReturn() and attr != .AlwaysTail) {
             _ = self.builder.buildUnreachable();
```

**Comment by wrongnull** [comment]
@Vexu got expected behavior on version ```0.12.0-dev.6053+d31fb9af9```

**Comment by Vexu** [comment]
Fixed by #16574
